src/components/marker-chart/index.js

```diff
@@ -80,35 +87,36 @@ class MarkerChart extends React.PureComponent<Props> {
         </div>
         <MarkerChartCanvas
           key={threadIndex}
-          // ChartViewport props
-          isRowExpanded={isRowExpanded}
-          isSelected={isSelected}
-          timeRange={timeRange}
-          maxViewportHeight={maxViewportHeight}
-          maximumZoom={this.getMaximumZoom()}
-          selection={selection}
-          updateProfileSelection={updateProfileSelection}
-          viewportNeedsUpdate={viewportNeedsUpdate}
-          // MarkerChartCanvas props
-          rangeStart={timeRange.start}
-          rangeEnd={timeRange.end}
-          markerTimingRows={markerTimingRows}
-          maxMarkerRows={maxMarkerRows}
-          markers={markers}
-          rowHeight={ROW_HEIGHT}
+          viewportProps={{
+            timeRange: timeRange,
+            maxViewportHeight: maxViewportHeight,
+            maximumZoom: this.getMaximumZoom(),
+            selection: selection,
+            viewportNeedsUpdate: viewportNeedsUpdate,
+          }}
+          chartProps={{
+            rangeStart: timeRange.start,
+            rangeEnd: timeRange.end,
+            markerTimingRows: markerTimingRows,
+            markers: markers,
+            rowHeight: ROW_HEIGHT,
+            updateProfileSelection: updateProfileSelection,
+          }}
         />
       </div>
     );
   }
 }
 
-function viewportNeedsUpdate(prevProps, newProps) {
+function viewportNeedsUpdate(
+  prevProps: MarkerChartCanvasOwnProps,
+  newProps: MarkerChartCanvasOwnProps
```

Comment:
Flow was not correctly finding the generic type. I figured out how to get a function to implement a given type, and correctly infer things.

```diff
diff --git a/src/components/stack-chart/index.js b/src/components/stack-chart/index.js
index 4349e917..9842a89d 100644
--- a/src/components/stack-chart/index.js
+++ b/src/components/stack-chart/index.js
@@ -146,8 +146,8 @@ export default explicitConnect(options);
 
 // Save an allocation by passing in the raw ChartProps.
 function viewportNeedsUpdate(
-  prevProps: StackChartCanvasOwnProps,
-  newProps: StackChartCanvasOwnProps
+  prevProps: { asdf: string },
+  newProps: { asdf: string }
 ) {
-  return prevProps.stackTimingByDepth !== newProps.stackTimingByDepth;
+  return prevProps.asdf !== newProps.asdf;
 }
```

```
e with. See: src/components/stack-chart/Canvas.js:331
            150:   newProps: { asdf: string }
                             ^^^^^^^^^^^^^^^^ object type
              Property `asdf` is incompatible:
                150:   newProps: { asdf: string }
                                 ^^^^^^^^^^^^^^^^ property `asdf`. Property not found in
                331: export default (withChartViewport: WithChartViewport<OwnProps, Props>)(
                                                                          ^^^^^^^^ OwnProps. See: src/components/
```

Reply:
Thanks, this works now :)

Now I still think we don't need the exact type, but just `{ +markerTimingRows: MarkerTimingRows }` (and now Flow will actually check this is correct thanks to your fix, I checked!):

This works and makes sense:
```javascript
function viewportNeedsUpdate(
  prevProps: { +markerTimingRows: MarkerTimingRows },
  newProps: { +markerTimingRows: MarkerTimingRows }
) {
  return prevProps.markerTimingRows !== newProps.markerTimingRows;
}
```

This makes more sense because this is the type we want in this function: an object with a readonly property named markerTimingRows. Flow ensures we get it.

And then you don't need to export OwnProps from Canvas, and I'm quite sure that if you don't need to split OwnProps from the viewport-injected props, things get also simpler in `shared/Viewport.js`. Although I'm fine if you don't want to refine this now.

So for this PR I think it's enough to:
* replace the type for viewportNeedsUpdate as explained abve
* stop exporting OwnProps from Canvas
and leave the rest for a later work (common with the simplification of the type of Viewport.js).

(actually the same in stack-chart/index.js:
```javascript
function viewportNeedsUpdate(
  prevProps: { +stackTimingByDepth: StackTimingByDepth },
  newProps: { +stackTimingByDepth: StackTimingByDepth }
) {
  return prevProps.stackTimingByDepth !== newProps.stackTimingByDepth;
}
```